src/crimson/osd/object_context.h

```diff
@@ -84,33 +85,8 @@ class ObjectContext : public Blocker,
   }
 
 private:
-  RWState rwstate;
-  seastar::shared_mutex wait_queue;
-  std::optional<seastar::shared_promise<>> wake;
-
-  template <typename F>
-  seastar::future<> with_queue(F &&f) {
-    return wait_queue.lock().then([this, f=std::move(f)] {
-      ceph_assert(!wake);
-      return seastar::repeat([this, f=std::move(f)]() {
-	if (f()) {
-	  wait_queue.unlock();
-	  return seastar::make_ready_future<seastar::stop_iteration>(
-	    seastar::stop_iteration::yes);
-	} else {
-	  rwstate.inc_waiters();
-	  wake = seastar::shared_promise<>();
-	  return wake->get_shared_future().then([this, f=std::move(f)] {
-	    wake = std::nullopt;
-	    rwstate.dec_waiters(1);
-	    return seastar::make_ready_future<seastar::stop_iteration>(
-	      seastar::stop_iteration::no);
-	  });
-	}
-      });
-    });
-  }
-
+  tri_mutex rwlock;
+  bool recovery_read_marker = false;
```

Comment:
Hmm, looks the sole purpose of this flag is to `assert()` in `drop_recovery_read()`. If that's correct, then [`NDEBUG`](https://stackoverflow.com/a/29253284) could be handy. On the other hand, guards aren't elegant...

Reply:
no, please see https://github.com/ceph/ceph/pull/37131/files#diff-54370af98a6f12723934c4dff4e882e6R93 .

Reply:
Ah, sure. Since aeb20dd2f5d757d1aa377022977d2277f4eac8ad it's necessary also for release builds.